src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.cs

```diff
@@ -650,11 +645,99 @@ public static IHttpClientBuilder ConfigureAdditionalHttpMessageHandlers(this IHt
             return builder;
         }
 
+        public static IHttpClientBuilder AddAsKeyed(this IHttpClientBuilder builder, ServiceLifetime lifetime = ServiceLifetime.Scoped)
+        {
+            ThrowHelper.ThrowIfNull(builder);
+
+            string? name = builder.Name;
+            IServiceCollection services = builder.Services;
+            HttpClientMappingRegistry registry = GetMappingRegistry(services);
+
+            UpdateEmptyNameHttpClient(services, registry);
+
+            if (name == null)
+            {
+                registry.DefaultKeyedLifetime?.RemoveRegistration(services);
+
+                registry.DefaultKeyedLifetime = new HttpClientKeyedLifetime(lifetime);
+                registry.DefaultKeyedLifetime.AddRegistration(services);
+            }
+            else
+            {
+                if (registry.KeyedLifetimeMap.TryGetValue(name, out HttpClientKeyedLifetime? clientLifetime))
+                {
+                    clientLifetime.RemoveRegistration(services);
+                }
+
+                clientLifetime = new HttpClientKeyedLifetime(name, lifetime);
+                registry.KeyedLifetimeMap[name] = clientLifetime;
+                clientLifetime.AddRegistration(services);
+            }
+
+            return builder;
+        }
+
+        public static IHttpClientBuilder RemoveAsKeyed(this IHttpClientBuilder builder)
+        {
+            ThrowHelper.ThrowIfNull(builder);
+
+            string? name = builder.Name;
+            IServiceCollection services = builder.Services;
+            HttpClientMappingRegistry registry = GetMappingRegistry(services);
+
+            UpdateEmptyNameHttpClient(services, registry);
+
+            if (name == null)
+            {
+                registry.DefaultKeyedLifetime?.RemoveRegistration(services);
+                registry.DefaultKeyedLifetime = HttpClientKeyedLifetime.Disabled;
+            }
+            else
+            {
+                if (registry.KeyedLifetimeMap.TryGetValue(name, out HttpClientKeyedLifetime? clientLifetime))
+                {
+                    clientLifetime.RemoveRegistration(services);
+                }
+                registry.KeyedLifetimeMap[name] = HttpClientKeyedLifetime.Disabled;
+            }
+
+            return builder;
+        }
+
+        // workaround for https://github.com/dotnet/runtime/issues/102654
+        private static void UpdateEmptyNameHttpClient(IServiceCollection services, HttpClientMappingRegistry registry)
+        {
+            if (registry.EmptyNameHttpClientDescriptor is not null)
+            {
+                bool removed = services.Remove(registry.EmptyNameHttpClientDescriptor);
```

Comment:
We do need it. Or at the very least we need to know that there _was_ an empty-name client descriptor that we've removed.

The code in the "parameterless" `AddHttpClient()` uses "TryAdd" for all the stuff it's adding, to make sure that we don't add more than one descriptor for every core HttpClientFactory-supporting service -- because this "parameterless" overload is being used like "ensure initialized" by all other `AddHttpClient(...)` methods.

So that would mean that if we've already "transformed" the `EmptyNameHttpClientDescriptor` into a keyed one, we still should not add a _new_ non-keyed descriptor upon the next `AddHttpClient(...)` call -- but given that it's not present in the service collection anymore, the ordinary `TryAdd...` would have ended up adding it. That's why its still needed -- for the check in `HttpClientFactoryServiceCollectionExtensions.TryAddEmptyNameHttpClient` inside `AddHttpClient()` (that can be called multiple times in any order with `AddAsKeyed`/`RemoveAsKeyed` calls)

Reply:
As discussed offline -- I'll add some comments with the explanations to the code in a follow-up PR (unless there would be some changes required within this one)